### PR TITLE
[refactor] make os stateful tensor

### PR DESCRIPTION
**Pull request by ver217**
related to #562 
Stateful OS tensors making them be managed dynamically in CPU-GPU space.

### CODE REVIEW

File: colossalai/zero/sharded_optim/sharded_optim_v2.py [hunk 1]
```diff
@@ -139,6 +139,8 @@ def __init__(self,
         if self._use_memory_tracer:
             GLOBAL_MODEL_DATA_TRACER.register_optimizer(self)
 
+        self._state: Dict[Parameter, Dict[str, StatefulTensor]] = {}
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:142]: will step() uses self._state?

File: colossalai/zero/sharded_optim/sharded_optim_v2.py [hunk 2]
```diff
@@ -358,3 +362,28 @@ def _copy_master_param_to_param_fp16(self):
...
+    def _set_optim_states_before_step(self):
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                # We don't need to set
+                # self.optim.state[p][k] = self._state[p][k].payload
+                # since payload is moved over devices inline
+                # self.optim.state[p][k]'s tensor id won't change
+                # the only change is its data pointer
+                for optim_state in self._state[p].values():
+                    optim_state.trans_state(TensorState.COMPUTE)
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: In this way, you set all OS tensor to compute?
You should set a tensor to COMPUTE the exact time it is about to be computed.

File: colossalai/zero/sharded_optim/sharded_optim_v2.py (hunk 1 above)
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:142]: It will use its payload.

File: colossalai/zero/sharded_optim/sharded_optim_v2.py (hunk 2 above)
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: We are unable to do this unless we use a custom optimizer.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: That is to say, users cannot use torch.optim.SGD, torch.optim.Adam anymore.

Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: I am afraid so. If you use the stateful tensor, then you have to use our stateful optimizer.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: I discuss with @1SAA . We both think that OS is not suitable for dynamic memory layout. Static layout like now maybe more  efficient.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: Move OS dynamically meybe not reduce computation time, however, it increases data movement time. Besides, it won't save momery usage.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: And we have to make sure OS, fp32 weight and gradient are on the same device for each parameter. That is to say, our eviction algorithm has to identify those stateful tensors are "linked" to which parameter. Although they are all "tensors", but they are not "independent" tensors.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: I think dynamic memory layout is suitable for fp16 params. OS, fp32 weight and gradient can be moved once and stay on the device.

Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:375]: OK, we make param fp16 stateful tensor first.
